bevy_lint/src/lints/suspicious/unit_in_bundle.rs

```diff
@@ -0,0 +1,415 @@
+            // won't be in `args`, making the two slices two different lengths.
+            let fn_arg_types = &fn_arg_types(cx, fn_id)[1..];
+
+            (fn_id, args, fn_arg_types)
+        } else if let ExprKind::Call(_, fn_args) = expr.kind {
+            let Some(fn_id) = fn_def_id(cx, expr) else {
+                // This will be `None` if the function is a local closure. Since closures
+                // cannot have generic parameters, they cannot take bundles as an input, so we
+                // can skip them.
+                return;
+            };
+
+            let fn_arg_types = fn_arg_types(cx, fn_id);
+
+            (fn_id, fn_args, fn_arg_types)
+        } else {
+            return;
+        };
+
+        debug_assert_eq!(fn_args.len(), fn_arg_types.len());
+
+        let typeck_results = cx.typeck_results();
+
+        for bundle_expr in filter_bundle_args(cx, fn_id, fn_args, fn_arg_types) {
+            let bundle_ty = typeck_results.expr_ty(bundle_expr);
+
+            for tuple_path in find_units_in_tuple(bundle_ty) {
+                let unit_expr = tuple_path.into_expr(bundle_expr);
+
+                span_lint_hir_and_then(
+                    cx,
+                    UNIT_IN_BUNDLE,
+                    unit_expr.hir_id,
+                    unit_expr.span,
+                    UNIT_IN_BUNDLE.desc,
+                    |diag| {
+                        diag.note("units `()` are not `Component`s and will be skipped");
+                    },
+                );
+            }
+        }
+    }
+}
+
+/// Returns the arguments of a method call that are intended to be `Bundle`s.
+///
+/// `fn_id` should be the definition of the function itself, and `args` should be the arguments
+/// passed to the function.
+fn filter_bundle_args<'tcx>(
+    cx: &LateContext<'tcx>,
+    fn_id: DefId,
+    fn_args: &'tcx [Expr<'tcx>],
+    fn_arg_types: &[Ty<'tcx>],
+) -> impl Iterator<Item = &'tcx Expr<'tcx>> {
+    let bundle_bounded_generics: Vec<Ty<'_>> = bundle_bounded_generics(cx, fn_id);
+
+    // Only yield arguments whose types are generic parameters that require the `Bundle` trait.
+    fn_arg_types
+        .iter()
+        .enumerate()
+        .filter(move |(_, arg)| bundle_bounded_generics.contains(arg))
+        .map(|(i, _)| &fn_args[i])
+}
+
+/// Returns a list of types corresponding to the inputs of a function.
+///
+/// Notably, the returned types are not instantiated. Generic parameters will be preserved and not
+/// filled in with actual types.
+///
+/// # Example
+///
+/// Running this function on the [`DefId`] of `foo()` will return `[usize, bool]`, while `bar()`
+/// will return `[T, usize]`.
+///
+/// ```
+/// fn foo(a: usize, b: bool) {}
+/// fn bar<T: Bundle>(bundle: T, size: usize) {}
+/// ```
+fn fn_arg_types<'tcx>(cx: &LateContext<'tcx>, fn_id: DefId) -> &'tcx [Ty<'tcx>] {
+    cx.tcx
+        .fn_sig(fn_id)
+        .instantiate_identity()
+        .inputs()
+        .skip_binder()
```

Comment:
We want to use `instantiate_identity()` and `skip_binder()` here because we want generic parameters to remain. If we instantiate it, the generic parameters will be replaced with actual concrete types, which isn't desired!